Add error handling for s3 client errors

ctaymor opened the pull request:
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/ctaymor/mlflow/pull/17121?quickstart=1)

#### Install mlflow from this PR

```
# mlflow
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/17121/merge
# mlflow-skinny
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/17121/merge#subdirectory=libs/skinny
```

For Databricks, use the following command:

```
%sh curl -LsSf https://raw.githubusercontent.com/mlflow/mlflow/HEAD/dev/install-skinny.sh | sh -s pull/17121/merge
```

</p>
</details>

### Related Issues/PRs

Resolve #17040

### What changes are proposed in this pull request?

<!-- Please fill in changes proposed in this PR. -->

### How is this PR tested?

- [ ] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests

*Manual testing:*
Run the local server, using a real s3 bucket, and fake access keys.
```bash
export AWS_ACCESS_KEY_ID=fake_key_id
export AWS_SECRET_ACCESS_KEY=fake_secret_key
export AWS_DEFAULT_REGION=us-east-1

mlflow server \
  --backend-store-uri file:///Users/<USERNAME>/workspace/mlflow/mlruns \
  --artifacts-destination s3://test-s3-mlflow-ct-08-05/artifacts \
  --host 127.0.0.1 \
  --port 5000
```
In a separate window, run the test-bug.sh code:
```bash
#!/usr/bin/env bash

set -eux

python -c "
import mlflow
import os

current_dir = os.getcwd()
mlflow.set_tracking_uri('http://127.0.0.1:5000')  # Point to the server

with mlflow.start_run() as run:
    mlflow.log_param('test', 'value')
    print(f'Created run with ID: {run.info.run_id}')

    with open('run_id_s3.txt', 'w') as f:
        f.write(run.info.run_id)
"

# Test with the new run ID
RUN_ID=$(cat run_id_s3.txt)
curl -X GET "http://127.0.0.1:5000/api/2.0/mlflow/artifacts/list?run_id=$RUN_ID" -H "accept: application/json" -v
```
Before the change, you should get a 500. After the change, you will get a 401 Unauthorized.


### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

### Release Notes

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Fixes a bug where list_artifacts was throwing a 500 error, when Mlflow receives a 4xx error from S3. The list_artifacts endpoint backed by S3 will now appropriately surface the 4xx error from S3.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows
- [ ] `area/gateway`: MLflow AI Gateway client APIs, server, and third-party integrations
- [ ] `area/prompts`: MLflow prompt engineering features, prompt templates, and prompt management
- [ ] `area/tracing`: MLflow Tracing features, tracing APIs, and LLM tracing functionality
- [ ] `area/projects`: MLproject format, project running backends
- [x] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

<!-- patch -->

- [x] Yes (this PR will be cherry-picked and included in the next patch release)
- [ ] No (this PR will be included in the next minor release)

ctaymor committed:
Add error handling for s3 client errors

Signed-off-by: Caroline Taymor <ctaymor@gmail.com>

ctaymor commented:
This PR resolves #17040, and keeps to that scope.
I think the other actions in the S3 Artifact Repository will have the same bug, and it may be worth making follow up issues to address error handling more broadly in the S3 Artifact Repository. If that is of interest, I would be happy to do follow up work on that.

serena-ruan commented on mlflow/server/handlers.py:
```diff
@@ -1151,7 +1151,6 @@ def list_artifacts_impl(request_message):
     return response_message
 
 
-@catch_mlflow_exception
```
Could we revert this change?

ctaymor committed:
Extract s3 pagination for list artifacts to helper

This makes it clearer where the ClientError could occur.
Note that because botocore paginates lazily, and the call does not
happen until the pagination is iterated over, the helper must yield the
results rather than returning them.

Signed-off-by: Caroline Taymor <ctaymor@gmail.com>

ctaymor commented on mlflow/store/artifact/s3_artifact_repo.py:
```diff
@@ -309,37 +323,49 @@ def list_artifacts(self, path=None):
+            prefix = dest_path + "/" if dest_path else ""
+            s3_client = self._get_s3_client()
+            paginator = s3_client.get_paginator("list_objects_v2")
+            results = paginator.paginate(Bucket=bucket, Prefix=prefix, Delimiter="/")
```
Yes, definitely! Good idea. I will make that change.

ctaymor committed:
Extract s3 uri parsing to a helper method

This change makes it a little easier to see that the pagination, and
handling of the results from pagination, are the heart of
list_artifacts.

Signed-off-by: Caroline Taymor <ctaymor@gmail.com>

serena-ruan reviewed (approved):
Overall LGTM! Let's revert unnecessary changes :) Thanks for the contribution!!

ctaymor committed:
Revert "Extract s3 uri parsing to a helper method"

This change was not necessary, it was a refactor to make the code
clearer to me. However, I'm happy to revert it by request from the
reviewer.
This reverts commit 3002689e2613949a180544127a487721e8f1693d.

Signed-off-by: Caroline Taymor <ctaymor@gmail.com>
